# Lab | Data Structuring and Combining Data

## Challenge 1: Combining & Cleaning Data

In this challenge, we will be working with the customer data from an insurance company, as we did in the two previous labs. The data can be found here:
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

But this time, we got new data, which can be found in the following 2 CSV files located at the links below.

- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file2.csv
- https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file3.csv

Note that you'll need to clean and format the new data.

Observation:
- One option is to first combine the three datasets and then apply the cleaning function to the new combined dataset
- Another option would be to read the clean file you saved in the previous lab, and just clean the two new files and concatenate the three clean datasets

In [1]:
import pandas as pd

# Read the datasets into pandas dataframes
df1 = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv")
df2 = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file2.csv")
df3 = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file3.csv")

#display(df1.head())
#display(df2.head())
#display(df3.head())

In [2]:
# Concatenate the three dataframes vertically
combined_df = pd.concat([df1, df2, df3], axis=0, ignore_index=True)

#display(combined_df)
#display("Shape of combined dataframe:", combined_df.shape)
#display(combined_df)

# check if combined_df shape = 1+2+3 
#display(df1.shape)
#display(df2.shape)
#display(df3.shape)

In [3]:
# Clean the combined df 
def clean_combined_date(combined_df):
    # Replace missing numeric values with the median
    numeric_cols = combined_df.select_dtypes(include=['number']).columns
    for col in numeric_cols:
        median_value = combined_df[col].median()
        combined_df[col].fillna(median_value, inplace=True)
    
    # Replace missing categorical values with 'unknown'
    categorical_cols = combined_df.select_dtypes(include=['object']).columns
    for col in categorical_cols:
       combined_df[col].fillna('unknown', inplace=True)
    
    # Combine 'st' and 'state' columns
    if 'st' in dcombined_df.columns and 'state' in combined_df.columns:
        combined_df['state_combined'] = combined_df['st'].combine_first(df['state'])
        combined_df.drop(['st', 'state'], axis=1, inplace=True)
    elif 'st' in df.columns:
        combined_df.rename(columns={'st': 'state_combined'}, inplace=True)
    elif 'state' in df.columns:
        combined_df.rename(columns={'state': 'state_combined'}, inplace=True)
    
    return(combined_df)

In [4]:
display(combined_df)
combined_df.to_csv('combined_df.csv', index=False)

Customer          ST GENDER             Education  \
0      RB50392  Washington    NaN                Master   
1      QZ44356     Arizona      F              Bachelor   
2      AI49188      Nevada      F              Bachelor   
3      WW63253  California      M              Bachelor   
4      GA49547  Washington      M  High School or Below   
...        ...         ...    ...                   ...   
12069  LA72316         NaN    NaN              Bachelor   
12070  PK87824         NaN    NaN               College   
12071  TD14365         NaN    NaN              Bachelor   
12072  UP19263         NaN    NaN               College   
12073  Y167826         NaN    NaN               College   

      Customer Lifetime Value   Income  Monthly Premium Auto  \
0                         NaN      0.0                1000.0   
1                  697953.59%      0.0                  94.0   
2                 1288743.17%  48767.0                 108.0   
3                  764586.18%      0.0                 106.0   
4                  536307.65%  36357.0                  68.0   
...                       ...      ...                   ...   
12069             23405.98798  71941.0                  73.0   
12070             3096.511217  21604.0                  79.0   
12071             8163.890428      0.0                  85.0   
12072             7524.442436  21941.0                  96.0   
12073             2611.836866      0.0                  77.0   

      Number of Open Complaints     Policy Type  Vehicle Class  \
0                        1/0/00   Personal Auto  Four-Door Car   
1                        1/0/00   Personal Auto  Four-Door Car   
2                        1/0/00   Personal Auto   Two-Door Car   
3                        1/0/00  Corporate Auto            SUV   
4                        1/0/00   Personal Auto  Four-Door Car   
...                         ...             ...            ...   
12069                         0   Personal Auto  Four-Door Car   
12070                         0  Corporate Auto  Four-Door Car   
12071                         3  Corporate Auto  Four-Door Car   
12072                         0   Personal Auto  Four-Door Car   
12073                         0  Corporate Auto   Two-Door Car   

       Total Claim Amount       State Gender  
0                2.704934         NaN    NaN  
1             1131.464935         NaN    NaN  
2              566.472247         NaN    NaN  
3              529.881344         NaN    NaN  
4               17.269323         NaN    NaN  
...                   ...         ...    ...  
12069          198.234764  California      M  
12070          379.200000  California      F  
12071          790.784983  California      M  
12072          691.200000  California      M  
12073          369.600000  California      M  

[12074 rows x 13 columns]

# Challenge 2: Structuring Data

In this challenge, we will continue to work with customer data from an insurance company, but we will use a dataset with more columns, called marketing_customer_analysis.csv, which can be found at the following link:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis_clean.csv

This dataset contains information such as customer demographics, policy details, vehicle information, and the customer's response to the last marketing campaign. Our goal is to explore and analyze this data by performing data cleaning, formatting, and structuring.

In [5]:
import pandas as pd

# Load the dataset
df = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/marketing_customer_analysis_clean.csv")

display(df.head(10))

unnamed:_0 customer       state  customer_lifetime_value response  \
0           0  DK49336     Arizona              4809.216960       No   
1           1  KX64629  California              2228.525238       No   
2           2  LZ68649  Washington             14947.917300       No   
3           3  XL78013      Oregon             22332.439460      Yes   
4           4  QA50777      Oregon              9025.067525       No   
5           5  HO81102  California              4745.181764       No   
6           6  IW72280  California              5035.035257       No   
7           7  IH64929  California              4956.247235       No   
8           8  FM55990  California              5989.773931      Yes   
9           9  QX31376      Oregon             17114.328150       No   

   coverage             education effective_to_date employmentstatus gender  \
0     Basic               College        2011-02-18         Employed      M   
1     Basic               College        2011-01-18       Unemployed      F   
2     Basic              Bachelor        2011-02-10         Employed      M   
3  Extended               College        2011-01-11         Employed      M   
4   Premium              Bachelor        2011-01-17    Medical Leave      F   
5     Basic  High School or Below        2011-02-14         Employed      M   
6     Basic                Doctor        2011-02-14         Employed      F   
7     Basic  High School or Below        2011-02-24         Employed      M   
8   Premium               College        2011-01-19         Employed      M   
9  Extended                Master        2011-01-04         Employed      F   

   ...  number_of_policies     policy_type        policy  renew_offer_type  \
0  ...                   9  Corporate Auto  Corporate L3            Offer3   
1  ...                   1   Personal Auto   Personal L3            Offer4   
2  ...                   2   Personal Auto   Personal L3            Offer3   
3  ...                   2  Corporate Auto  Corporate L3            Offer2   
4  ...                   7   Personal Auto   Personal L2            Offer1   
5  ...                   7   Personal Auto   Personal L3            Offer1   
6  ...                   4  Corporate Auto  Corporate L2            Offer2   
7  ...                   3   Personal Auto   Personal L2            Offer2   
8  ...                   1   Personal Auto   Personal L1            Offer2   
9  ...                   2  Corporate Auto  Corporate L3            Offer2   

   sales_channel  total_claim_amount  vehicle_class  vehicle_size  \
0          Agent          292.800000  Four-Door Car       Medsize   
1    Call Center          744.924331  Four-Door Car       Medsize   
2    Call Center          480.000000            SUV       Medsize   
3         Branch          484.013411  Four-Door Car       Medsize   
4         Branch          707.925645  Four-Door Car       Medsize   
5          Agent          292.800000  Four-Door Car       Medsize   
6         Branch          287.556107  Four-Door Car       Medsize   
7         Branch          247.402353   Two-Door Car       Medsize   
8         Branch          739.200000     Sports Car       Medsize   
9          Agent          408.000000  Four-Door Car         Small   

  vehicle_type month  
0            A     2  
1            A     1  
2            A     2  
3            A     1  
4            A     1  
5            A     2  
6            A     2  
7            A     2  
8            A     1  
9            A     1  

[10 rows x 27 columns]

In [6]:
# Clean the df
def clean_df(df):
    # Check and handle for missing values 
    if df.isnull().any().any():
        if 'customer' in df.columns and df['customer'].dtype in [float, int]:
            df['customer'].fillna(df['customer'].mean(), inplace=True)

    # Standardize column names
    df.columns = [col.lower().replace(' ', '_') for col in df.columns]

    # Format categorical data 
    categorical_columns = ['state', 'gender', 'education', 'location_code', 'marital_status', 'policy_type', 'policy', 'renew_offer_type', 'sales_channel', 'vehicle_class', 'vehicle_size']
    for col in categorical_columns:
        df[col] = df[col].astype('category')
    
    return df

1. You work at the marketing department and you want to know which sales channel brought the most sales in terms of total revenue. Using pivot, create a summary table showing the total revenue for each sales channel (branch, call center, web, and mail).
Round the total revenue to 2 decimal points.  Analyze the resulting table to draw insights.

2. Create a pivot table that shows the average customer lifetime value per gender and education level. Analyze the resulting table to draw insights.

In [18]:
# Pivot table to show total revenue for each sales channel
sales_channel_revenue=df.pivot_table(index='sales_channel', values='total_claim_amount', aggfunc='sum').round(2)
display(sales_channel_revenue)

# Pivot table to show average customer lifetime value per gender and education level
pd.pivot_table(df, index=['gender', 'education'], values='customer_lifetime_value', aggfunc='mean').round(2)

total_claim_amount
sales_channel                    
Agent                  1810226.82
Branch                 1301204.00
Call Center             926600.82
Web                     706600.04

customer_lifetime_value
gender education                                    
F      Bachelor                              7874.27
       College                               7748.82
       Doctor                                7328.51
       High School or Below                  8675.22
       Master                                8157.05
M      Bachelor                              7703.60
       College                               8052.46
       Doctor                                7415.33
       High School or Below                  8149.69
       Master                                8168.83

Analyze the resulting table to draw insights.
1. Agent brought the most sales in terms of total revenue, which is 1810226.82. 
2. Customers with a Bachelor's degree have the highest lifetime value for both females and males. Next are college and doctor.

## Bonus

You work at the customer service department and you want to know which months had the highest number of complaints by policy type category. Create a summary table showing the number of complaints by policy type and month.
Show it in a long format table.

*In data analysis, a long format table is a way of structuring data in which each observation or measurement is stored in a separate row of the table. The key characteristic of a long format table is that each column represents a single variable, and each row represents a single observation of that variable.*

*More information about long and wide format tables here: https://www.statology.org/long-vs-wide-data/*